functions/calcTajimaD.txt

```diff
@@ -0,0 +1,65 @@
+// Written by Nick Bailey, submitted on 2024-06-13
+// Currently affiliated with CNRS and University of Lyon 1
+// This code is provided in the public domain free of restriction, please credit SLiM-Extras (https://github.com/MesserLab/SLiM-Extras) and myself, Nick Bailey, if you use it
+
+// calculate neutrality test based on the site frequency spectrum, Tajima's D
+// calculations originally in Tajima 1989
+function (float)calcTajimaD(object<Genome> genomes, [No<Mutation> muts = NULL], [Ni$ start = NULL], [Ni$ end = NULL])
+// conduct various checks for proper usage copied straight from calcWattersonTheta()
+{ if (genomes.length() == 0)
+		stop("ERROR (calcTajimaD()): genomes must be non-empty.");
+	if (community.allSpecies.length() > 1)
+	{
+		species = unique(genomes.individual.subpopulation.species, preserveOrder=F);
+		if (species.length() != 1)
+			stop("ERROR (calcTajimaD()): genomes must all belong to the same species.");
+		if (!isNULL(muts))
+			if (!all(muts.mutationType.species == species))
+				stop("ERROR (calcTajimaD()): muts must all belong to the same species as genomes.");
+	}
+	else
+	{
+		species = community.allSpecies;
+	}
+	
+	if (isNULL(muts))
+		muts = species.mutations;
+	
+	// handle windowing
+	if (!isNULL(start) & !isNULL(end))
+	{
+		if (start > end)
+			stop("ERROR (calcTajimaD()): start must be less than or equal to end.");
+		mpos = muts.position;
+		muts = muts[(mpos >= start) & (mpos <= end)];
+		length = end - start + 1;
+	}
+	else if (!isNULL(start) | !isNULL(end))
+	{
+		stop("ERROR (calcTajimaD()): start and end must both be NULL or both be non-NULL.");
+	}
+
+	// narrow down to the mutations that are actually present in the genomes and aren't fixed
+	p = genomes.mutationFrequenciesInGenomes(muts);
+	muts = muts[(p != 0.0) & (p != 1.0)];
+	
+	// do the calculation
+	// Pi and Watterson's theta functions divide by sequence length so this must be undone in Tajima's D
+	// Sequence length is constant (i.e. no missing data or indels) so this can be applied equally over both metrics
+	diff = (calcPiTheta(genomes) - calcWattersonsTheta(genomes)) * (species.chromosome.lastPosition + 1);
```

Comment:
Shouldn't `calcPiTheta( )` have `start`, `end`, and `muts` arguments passed in as well?

Reply:
That makes sense but I don't know how best to do this, as those should be optional. I think I need to set up a condition if the user does enter that info? As it is now, it works fine genome-wide but entering starts and ends to Tajima's D does not change the pi and WT calculations so I do need to fix this somehow, or just disallow the extra options.

Reply:
Can't you just do
```suggestion
	diff = (calcPiTheta(genomes, muts, start, end) - calcWattersonsTheta(genomes, muts, start, end)) * (species.chromosome.lastPosition + 1);
```
... ?

Reply:
Yep, that works, thanks! Another thing I'm not entirely sure what I thinking. I think I was assuming if I put muts, starts, and ends into the functions it would return errors when the user never actually defined those, and I figured I'd try it later but never got to it. I see that the kinds of conditions I mentioned would be necessary are already in the function (e.g. if (isNULL(muts)) muts = species.mutations;), so it's already handling those cases and defining the variables appropriately. Running it this way seems to work as expected.